karapace/kafka_rest_apis/__init__.py

```diff
@@ -279,24 +279,31 @@ def get_topic_config(self, topic: str) -> dict:
             return self.admin_client.get_topic_config(topic)
 
     def cluster_metadata(self, topics: Optional[List[str]] = None) -> dict:
-        if self._metadata_birth is None or time.time() - self._metadata_birth > self.metadata_max_age:
-            self._cluster_metadata = None
-        if not self._cluster_metadata:
-            self._metadata_birth = time.time()
-            with self.admin_lock:
-                try:
-                    self._cluster_metadata = self.admin_client.cluster_metadata(topics)
-                except NodeNotReadyError:
-                    self.log.exception("Could not refresh cluster metadata")
-                    self.r(
-                        body={
-                            "message": "Kafka node not ready",
-                            "code": RESTErrorCodes.HTTP_INTERNAL_SERVER_ERROR.value,
-                        },
-                        content_type="application/json",
-                        status=HTTPStatus.INTERNAL_SERVER_ERROR,
-                    )
-        return copy.deepcopy(self._cluster_metadata)
+        with self.admin_lock:
+            if self._metadata_birth is None or time.monotonic() - self._metadata_birth > self.metadata_max_age:
+                self._cluster_metadata = None
+
+            if self._cluster_metadata and topics is None:
```

Comment:
question: why is the cache being skipped if the argument topics is provided? Is that intentional?

Reply:
I did it intentionally to always return only given topics. This code could be place for more refactoring, I agree, but for now I was trying to make it correct first.